File: PathoFroh/src/main/java/com/patho/main/action/dialog/settings/slide/StainingEditDialog.java
package com.patho.main.action.dialog.settings.slide;

import com.patho.main.action.dialog.AbstractDialog;
import com.patho.main.common.Dialog;
import com.patho.main.model.StainingPrototype;
import com.patho.main.model.StainingPrototypeDetails;
import com.patho.main.service.impl.SpringContextBridge;
import com.patho.main.util.dialog.event.ReloadEvent;
import lombok.Getter;
import lombok.Setter;

import java.sql.Date;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
public class StainingEditDialog extends AbstractDialog {

    private boolean newStaining;

    private StainingPrototype stainingPrototype;

    private List<StainingPrototypeDetails> removeDetails;

    public StainingEditDialog initAndPrepareBean() {
        return initAndPrepareBean(new StainingPrototype());
    }

    public StainingEditDialog initAndPrepareBean(StainingPrototype staining) {
        if (initBean(staining))
            prepareDialog();
        return this;
    }

    public boolean initBean(StainingPrototype staining) {

        // reloading stating with batch details
        if (staining.getId() != 0) {
            setStainingPrototype(
                    SpringContextBridge.services().getStainingPrototypeRepository().findOptionalByIdAndInitilize(staining.getId(), true).get());
            setNewStaining(false);
        } else {
            setNewStaining(true);
            setStainingPrototype(staining);
        }

        setRemoveDetails(new ArrayList<StainingPrototypeDetails>());

        return super.initBean(task, Dialog.SETTINGS_STAINING_EDIT);
    }

    public void addBatch() {
        StainingPrototypeDetails newBatch = new StainingPrototypeDetails(stainingPrototype);
        newBatch.setDeliveryDate(new Date(System.currentTimeMillis()));
        getStainingPrototype().getBatchDetails().add(0, newBatch);
    }

    public void removeBatch(StainingPrototypeDetails stainingPrototypeDetails) {
        stainingPrototype.getBatchDetails().remove(stainingPrototypeDetails);

        // delete if user clicks save
        if (stainingPrototypeDetails.getId() != 0) {
            getRemoveDetails().add(stainingPrototypeDetails);
        }
    }

    public void cloneBatch(StainingPrototypeDetails stainingPrototypeDetails) {
        try {
            StainingPrototypeDetails clone = stainingPrototypeDetails.clone();
            clone.setId(0);
            stainingPrototype.getBatchDetails().add(0, clone);
        } catch (CloneNotSupportedException e) {
        }
    }

    public void saveAndHide() {
        save();
        hideDialog(new ReloadEvent());
    }

    private void save() {
        SpringContextBridge.services().getStainingPrototypeService().addOrUpdate(getStainingPrototype(), getRemoveDetails());
    }
}

File: PathoFroh/src/main/kotlin/com/patho/main/dialog/settings/staining/StainingEditDialog.kt
package com.patho.main.dialog.settings.staining

import com.patho.main.common.Dialog
import com.patho.main.dialog.AbstractTaskDialog
import com.patho.main.model.preset.StainingPrototype
import com.patho.main.model.preset.StainingPrototypeDetails
import com.patho.main.repository.jpa.StainingPrototypeRepository
import com.patho.main.service.impl.SpringContextBridge.Companion.services
import com.patho.main.util.dialog.event.ReloadEvent
import com.patho.main.util.exceptions.EntityNotFoundException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

/**
 * Dialog for creating or editing stainings
 */
@Component()
open class StainingEditDialog @Autowired constructor(
        private val stainingPrototypeRepository: StainingPrototypeRepository) : AbstractTaskDialog(Dialog.SETTINGS_STAINING_EDIT) {

    open lateinit var stainingPrototype: StainingPrototype

    val isNewStaining
        get() = stainingPrototype.id == 0L

    override fun initAndPrepareBean(): StainingEditDialog {
        return  initAndPrepareBean(StainingPrototype())
    }

    open fun initAndPrepareBean(prototype: StainingPrototype): StainingEditDialog {
        if (initBean(prototype))
            prepareDialog()
        return this
    }

    open fun initBean(prototype: StainingPrototype): Boolean {
        this.stainingPrototype = prototype

        if (!isNewStaining) {
            val oStainingPrototype = stainingPrototypeRepository.findOptionalByIdAndInitialize(this.stainingPrototype.id)

            if (!oStainingPrototype.isPresent)
                throw EntityNotFoundException()
            else
                this.stainingPrototype = oStainingPrototype.get()
        }

        return super.initBean()
    }

    /**
     * Create a new batch
     */
    open fun addBatch() {
        val newBatch = StainingPrototypeDetails()
        this.stainingPrototype.batchDetails.add(0, newBatch)
    }

    /**
     * Remove a batch
     */
    open fun removeBatch(stainingPrototypeDetails: StainingPrototypeDetails?) {
        this.stainingPrototype.batchDetails.remove(stainingPrototypeDetails)
    }

    /**
     * Clone a batch
     */
    open fun cloneBatch(stainingPrototypeDetails: StainingPrototypeDetails) {
        this.stainingPrototype.batchDetails.add(0, (stainingPrototypeDetails.clone() as StainingPrototypeDetails))
    }

    open fun saveAndHide() {
        save()
        hideDialog(ReloadEvent())
    }

    open fun save() {
        services().stainingPrototypeService.addOrUpdate(this.stainingPrototype)
    }

}